File: clients/typescript/test/smoke.mjs
// Runtime smoke test: drive the built REST transport against a live
// `pawrly console`. Run via `pnpm smoke` (needs the built binary + SDK).
import { PawrlyClient, PawrlyError, UnsupportedByTransport } from "../dist/index.js";

const port = process.env.PAWRLY_PORT;
if (!port) {
  console.error("PAWRLY_PORT not set");
  process.exit(1);
}

const client = new PawrlyClient({
  transport: "rest",
  baseUrl: `http://127.0.0.1:${port}`,
});

function fail(msg) {
  console.error("SMOKE FAIL:", msg);
  process.exit(1);
}

// Wait for the console to come up.
const deadline = Date.now() + 10_000;
for (;;) {
  try {
    if ((await client.health()).ok) break;
  } catch {
    /* not up yet */
  }
  if (Date.now() > deadline) fail("console never came up");
  await new Promise((r) => setTimeout(r, 100));
}

const h = await client.health();
console.log("health:", JSON.stringify(h));
if (!h.ok) fail("health not ok");

// query() -> streaming QueryHandle, then collect().
const res = await (await client.query("SELECT 1 AS n")).collect();
console.log("query:", JSON.stringify(res));
if (res.rowCount !== 1) fail(`expected rowCount 1, got ${res.rowCount}`);
if (res.columns[0] !== "n") fail(`expected column n, got ${res.columns}`);
if (res.rows[0]?.n !== 1) fail(`expected n=1, got ${JSON.stringify(res.rows[0])}`);

// Async iteration (NDJSON streaming under the hood).
let streamed = 0;
for await (const row of await client.query("SELECT 1 AS n")) {
  streamed += 1;
  if (row.n !== 1) fail(`stream row n=${JSON.stringify(row)}`);
}
if (streamed !== 1) fail(`expected 1 streamed row, got ${streamed}`);
console.log("streamed rows:", streamed);

// materialize a query result, then read it back.
const out = await client.materialize("smoke_mat", {
  kind: "query",
  sql: "SELECT 1 AS n",
});
console.log("materialize:", JSON.stringify(out));
if (out.name.table !== "smoke_mat") fail(`materialize name ${JSON.stringify(out.name)}`);
const back = await (await client.query("SELECT * FROM materialized.smoke_mat")).collect();
if (back.rowCount !== 1) fail(`materialized readback rowCount ${back.rowCount}`);
console.log("materialized readback rowCount:", back.rowCount);

// semanticQuery reaches /v1/query; an empty workspace (no models) errors —
// which proves the endpoint + request serialization + error mapping.
try {
  await client.semanticQuery({ measures: ["orders.revenue"] });
  fail("semanticQuery should have errored (no models)");
} catch (e) {
  if (!(e instanceof PawrlyError)) fail(`semanticQuery wrong error: ${e}`);
  console.log("semanticQuery -> PawrlyError:", e.code);
}

// shutdown is unsupported over REST.
try {
  await client.shutdown();
  fail("shutdown should have thrown");
} catch (e) {
  if (!(e instanceof UnsupportedByTransport) || e.code !== "PAWRLY_UNSUPPORTED") {
    fail(`wrong shutdown error: ${e}`);
  }
  console.log("shutdown -> UnsupportedByTransport (PAWRLY_UNSUPPORTED)");
}

console.log("SMOKE OK");

File: clients/python/test/smoke.py
"""Runtime smoke: drive the REST transport against a live `pawrly console`."""

import os
import sys
import time

sys.path.insert(0, os.path.join(os.path.dirname(__file__), "..", "src"))

import pawrly  # noqa: E402


def fail(msg: str) -> None:
    print("SMOKE FAIL:", msg)
    sys.exit(1)


port = os.environ.get("PAWRLY_PORT")
if not port:
    fail("PAWRLY_PORT not set")

c = pawrly.PawrlyClient.rest(f"http://127.0.0.1:{port}")

deadline = time.time() + 10
while True:
    try:
        if c.health().ok:
            break
    except Exception:
        pass
    if time.time() > deadline:
        fail("console never came up")
    time.sleep(0.1)

h = c.health()
print("health:", h)
if not h.ok:
    fail("health not ok")

# query() -> streaming QueryHandle, then collect().
res = c.query("SELECT 1 AS n").collect()
print("query:", res)
if res.row_count != 1 or res.rows[0].get("n") != 1:
    fail(f"unexpected query result {res}")

# streaming iteration (NDJSON under the hood).
streamed = list(c.query("SELECT 1 AS n"))
if len(streamed) != 1:
    fail(f"expected 1 streamed row, got {len(streamed)}")
print("streamed rows:", len(streamed))

# materialize a query result, then read it back.
out = c.materialize("smoke_mat_py", pawrly.MaterializeSpec(sql="SELECT 1 AS n"))
print("materialize:", out)
if out.name.get("table") != "smoke_mat_py":
    fail(f"materialize name {out.name}")
back = c.query("SELECT * FROM materialized.smoke_mat_py").collect()
if back.row_count != 1:
    fail(f"materialized readback row_count {back.row_count}")
print("materialized readback row_count:", back.row_count)

# semantic_query reaches /v1/query; an empty workspace (no models) errors —
# which proves the endpoint + request serialization + error mapping.
try:
    c.semantic_query(pawrly.SemanticQuery(measures=["orders.revenue"]))
    fail("semantic_query should have errored (no models)")
except pawrly.PawrlyError as e:
    print("semantic_query -> PawrlyError:", e.code)

# shutdown is unsupported over REST.
try:
    c.shutdown()
    fail("shutdown should have raised")
except pawrly.UnsupportedByTransport as e:
    if e.code != "PAWRLY_UNSUPPORTED":
        fail(f"wrong shutdown code {e.code}")
    print("shutdown -> UnsupportedByTransport (PAWRLY_UNSUPPORTED)")

print("PY SMOKE OK")
